### PR TITLE
Added a function to support data conversion if type is not supported by the taipy.gui.data module

### LINKED ISSUE #457: Tabular Data: support a callback to change unsupported data type to supported one
**What would that feature address**
display unsupported data type

***Description of the ideal solution***
If a dataframe holds a column with unsupported data, we could handle a user callback that would transform that column to a supported type

***Caveats***
performance ?

***Other options***
Raise an error (as of today)

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented

### PR DESCRIPTION
**Pull request by Ujj1225**
Resolves: #457 

So there was this issue where we thought of giving developer option to change unsupported data format to a supported type
For this I have made two functions in class Gui namely "on_invalid_data" and "set_on_invalid_data_callback"

The Second function "set_on_invalid_data_callback" accepts a callback as parameter and uses this function to transform data into supported format. If user doesn't provide a callback, it works as it used to. Nothing would change. 

Besides in data_accessor, I have modified _InvalidDataAccessor. For all methods like get_data, get_col_types, etc. It calls the first function "on_invalid_data" to transform it into a valid data type. If no callback present it will work as it used to work.

### CODE REVIEW

**Comment by Ujj1225** [comment]
@quest-bot loot #457

File: taipy/gui/data/data_accessor.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ class _DataAccessor(ABC):
     _WS_DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
 
     def __init__(self, gui: "Gui") -> None:
+
```
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: we have our own formatting rules, we kindly request you to respect them

File: taipy/gui/data/data_accessor.py [hunk 2]
```diff
@@ -66,34 +74,59 @@ def to_csv(self, var_name: str, value: t.Any) -> t.Optional[str]:
...
 
     def get_data(
-        self, var_name: str, value: t.Any, payload: t.Dict[str, t.Any], data_format: _DataFormat
+        self,
+        var_name: str,
+        value: t.Any,
+        payload: t.Dict[str, t.Any],
+        data_format: _DataFormat,
     ) -> t.Dict[str, t.Any]:
+        transformed_value = self._gui.on_invalid_data(value)
```
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: I think the callback should not be invoked here.
It should be invoked when evaluating the right `DataAccessor` for the data type.
this way, we can redirect to an existing `DataAccessor` and keep the `InvalidDataAccessor` for invalid data types.

File: taipy/gui/gui.py [hunk 3]
```diff
@@ -396,7 +396,28 @@ def __init__(
         if libraries is not None:
             for library in libraries:
                 Gui.add_library(library)
+    def set_on_invalid_data_callback(self, callback: t.Callable[[t.Any], t.Optional[t.Any]]):
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: on_invalid_data should be handled the same way as on_action

- [definition](https://github.com/Avaiga/taipy/blob/3297952c0e64bd10358fde8e63a3adf6aef4e613/taipy/gui/gui.py#L255)
- [assignment](https://github.com/Avaiga/taipy/blob/3297952c0e64bd10358fde8e63a3adf6aef4e613/taipy/gui/gui.py#L2583)

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Good work @Ujj1225 
Please, move the invocation of the callback
we need tests

**Comment by Ujj1225** [comment]
hey, @FredLL-Avaiga  
I am so sorry for this formatting issue (turns out my prettier mistakely formatted the code I'll make sure to get it reverted)
So i changed the function on_invalid_data's definition and assignment following the way on_action is defined and assigned. 
Besides I also did run some tests. I havent pushed the file. Its called test_unsupported.py
What scenarios did I test?
- No callback
- Callback that returns None
- Callback that transforms data
- Callback that raises an exception
Its results
image
Do i need to push this file as well? I created it inside gui_specific tests

File: taipy/gui/gui.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: please remove this function

File: taipy/gui/gui.py [hunk 4]
```diff
@@ -405,6 +412,30 @@ def __init__(
...
+        self.on_invalid_data = callback
+
+    def handle_invalid_data(self, value: t.Any) -> t.Optional[t.Any]:
+        """
+        Handles unsupported data by invoking the callback if available.
+        :param value: The unsupported data encountered.
+        :return: Transformed data or None if no transformation is possible.
+        """
+        try:
+            if self.on_invalid_data:
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: could use `_is_function(self.on_invalid_data)`

File: taipy/gui/data/data_accessor.py [hunk 5]
```diff
@@ -34,7 +34,11 @@ def get_supported_classes() -> t.List[t.Type]:
 
     @abstractmethod
     def get_data(
-        self, var_name: str, value: t.Any, payload: t.Dict[str, t.Any], data_format: _DataFormat
+        self,
```
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: please remove your formatting, it makes reading your changes complicated

File: taipy/gui/data/data_accessor.py [hunk 6]
```diff
@@ -132,17 +152,27 @@ def _register(self, cls: t.Type[_DataAccessor]) -> None:
...
 
-    def __get_instance(self, value: _TaipyData) -> _DataAccessor:  # type: ignore
+
+    def __get_instance(self, value: _TaipyData) -> _DataAccessor:
         value = value.get() if isinstance(value, _TaipyData) else value
         access = self.__access_4_type.get(type(value))
         if access is None:
             if value is not None:
+                transformed_value = self.__gui.handle_invalid_data(value)
+                if transformed_value is not None:
```
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: here we should check the availability of an accessor with the new type and return it.

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

**Comment by Ujj1225** [comment]
hey i wrote a test case for the _get_instance as well 
image

File: taipy/gui/data/data_accessor.py [hunk 7]
```diff
@@ -137,6 +137,12 @@ def __get_instance(self, value: _TaipyData) -> _DataAccessor:  # type: ignore
...
+                    transformed_access = self.__access_4_type.get(type(transformed_value))
+                    if transformed_access is not None:
+                        return transformed_access
+                    return self.__invalid_data_accessor
```
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: no need for this line, the user will get the warning

File: taipy/gui/gui.py [hunk 8]
```diff
@@ -406,6 +413,22 @@ def __init__(
...
                 Gui.add_library(library)
 
+    def handle_invalid_data(self, value: t.Any) -> t.Optional[t.Any]:
+        """
+        Handles unsupported data by invoking the callback if available.
+        :param value: The unsupported data encountered.
+        :return: Transformed data or None if no transformation is possible.
+        """
+        try:
+            if self.on_invalid_data is not None and _is_function(self.on_invalid_data):
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: _is_function covers None

File: taipy/gui/gui.py [hunk 9]
```diff
@@ -406,6 +413,22 @@ def __init__(
...
+                return self.on_invalid_data(value)
+            else:
+                _warn(f"Unsupported data type encountered: {type(value)}")
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: no need for a warning here

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
I don't see your tests ?

**Comment by Ujj1225** [comment]
> I don't see your tests ?

Hey i have pushed test featuring the method __get_instance of data_accessor. This is where we are actually using transformed value and getting the access type for the transformed value

File: tests/gui/gui_specific/test_get_instance.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
+from taipy.gui import Gui
+from taipy.gui.data.data_accessor import _DataAccessors, _InvalidDataAccessor
+
+
+def test__get_instance_with_valid_data(gui: Gui):
+    """Test if __get_instance returns the correct accessor for valid data."""
+    data_accessors = _DataAccessors(gui)
+
+    # Simulate valid data being passed
+    data_accessors._DataAccessors__access_4_type = {int: "valid_accessor"}  # type: ignore
```
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]: you could have register the type through Gui

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py:1]: I think you should test through the public API as much as possible 
ie use get_data() to test __get_instance()

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

**Comment by Ujj1225** [comment]
hey @FredLL-Avaiga 
I am struggling with the tests 
As you suggested i tried to interact with the private get_instance using get_data but seems like get_data expects an object of type _TaipyData
So direct passing of int or str gives me assertion error 
and if i create a mocktaipydata to mimic it then again i get an error saying 
Can't find Data Accessor for type <class 'tests.gui.gui_specific.test_get_instance.MockTaipyData'>

How do i write the tests so it checks my get_instance through public api. I ran into a lot of assertion issues. This is why i tried to access the get_instance directly. 
plus you said i could register the type through gui. Which function does this?

File: tests/gui/gui_specific/test_get_instance.py (hunk 10 above)
Comment (FredLL-Avaiga) [reply tests/gui/gui_specific/test_get_instance.py]: gui._get_accessor()._register(<your testing accessor>)

File: tests/gui/gui_specific/test_get_instance.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    result = data_accessors._DataAccessors__get_instance(123)  # type: ignore
+    assert result == "valid_accessor"  # Valid accessor should be returned
+
+
+def test__get_instance_with_invalid_data(gui: Gui):
+    """Test if __get_instance handles invalid data correctly."""
+    data_accessors = _DataAccessors(gui)
+
+    # No valid data accessor, should return None
+    gui.handle_invalid_data = lambda x: None  # type: ignore
```
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]: gui._get_accessor().get_data(var_name, newvalue, payload)

**Comment by FredLL-Avaiga** [comment]
any news ?
please update your branch

**Comment by FredLL-Avaiga** [comment]
Are you still working on this PR @Ujj1225  ?
If you are, please resolve the conflicts, update your branch and answer the questions
If you are not, please close the PR

**Comment by Ujj1225** [comment]
yes @FredLL-Avaiga  I am working on it I'll resolve the conflicts and update the branch soon!

**Comment by Ujj1225** [comment]
hey @FredLL-Avaiga  I have now swtiched to use of public api _get_data to check for the method _get_instance but still I am struggling with the use of _get_accessor()._register()

**Comment by FredLL-Avaiga** [comment]
your branch still have conflicts

File: taipy/gui/gui.py [hunk 12]
```diff
@@ -406,6 +413,21 @@ def __init__(
             for library in libraries:
                 Gui.add_library(library)
 
+    def handle_invalid_data(self, value: t.Any) -> t.Optional[t.Any]:
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: make this method private by prefixing it with _

File: tests/gui/gui_specific/test_get_instance.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+from taipy.gui.utils.types import _TaipyData
+from unittest.mock import Mock
+
+
+def mock_taipy_data(value):
+    """Helper to mock _TaipyData objects."""
+    mock_data = Mock(spec=_TaipyData)
+    mock_data.get.return_value = value
+    return mock_data
+
```
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]:
```
class MyDataAccessor(_DataAccessor):
    @staticmethod
    def get_supported_classes() -> t.List[t.Type]:
        return [int]
  def get_data(
        self, var_name: str, value: t.Any, payload: t.Dict[str, t.Any], data_format: _DataFormat
    ) -> t.Dict[str, t.Any]:
      return {"value": value}
 ...implement
```

File: tests/gui/gui_specific/test_get_instance.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+def test_get_data_with_valid_data(gui: Gui):
+    """Test if get_data() returns the correct accessor for valid data."""
+    data_accessors = _DataAccessors(gui)
+    data_accessors._DataAccessors__access_4_type = {int: Mock(get_data=lambda *args: "valid_data")}  # type: ignore
```
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]:
```
data_accessors._register(MyDataAccessor)
result= data_accessors.get_data("var_name", 100, {})
assert result.get("value") == 100
```

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
I began some test implementation
Can you check it out and finish it ?

File: tests/gui/gui_specific/test_get_instance.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        pass
+
+    def to_csv(self, var_name: str, value: t.Any) -> t.Optional[str]:
+        pass
+
+
+def test_get_data_with_valid_data(gui: Gui):
+    """Test if get_data() returns the correct accessor for valid data."""
+    data_accessors = _DataAccessors(gui)
+    data_accessors._DataAccessors__access_4_type = {int: Mock(get_data=lambda *args: "valid_data")}  # type: ignore
```
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]: why don't you use the _register method and the MyDataAccessor class ?

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
are there still some linter issue ?

**Comment by RymMichaut** [comment]
Congratulations @Ujj1225 on this merged PR
Could you please contact me at rym.michaut@taipy.io for the shipping?

File: tests/gui/gui_specific/test_get_instance.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    result = data_accessors.get_data("var_name", mock_taipy_data(123), {})
+
+    assert result == "valid_data"
+
+
+def test_get_data_with_invalid_data(gui: Gui):
+    """Test if get_data() handles invalid data correctly."""
+    data_accessors = _DataAccessors(gui)
+    gui._handle_invalid_data = lambda x: None  # type: ignore
```
Comment (FabienLelaquais) [inline tests/gui/gui_specific/test_get_instance.py]:
```suggestion
    gui._handle_invalid_data = lambda _: None
```

File: tests/gui/gui_specific/test_get_instance.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    result = data_accessors.get_data("var_name", mock_taipy_data, {})
+    print(f"Result from get_data: {result}")
+
+    assert result == {}, f"Expected {{}} but got {result}"
+
+
+def test_get_data_transformation_successful(gui: Gui):
+    """Test if get_data() transforms invalid data correctly when a callback is set."""
+    data_accessors = _DataAccessors(gui)
+    gui.handle_invalid_data = lambda x: 123  # type: ignore
```
Comment (FabienLelaquais) [inline tests/gui/gui_specific/test_get_instance.py]:
```suggestion
    gui.handle_invalid_data = lambda _: 123  # type: ignore
```

**Comment by FabienLelaquais** [comment]
So I have a few remarks after I ultimately got the time to look at this.
First of all well done, @Ujj1225. What you propose does the job, and instructions were followed. And congrats for the tests!
We will integrate this into Taipy BUT only after I'm happy with the proposed API, which I'm not. But hey, nobody could have thought, since I never opened my mouth.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

Here are the reasons why I'm reluctant to merge as-is:

- Really, this thing is *not* a callback. In all of Taipy's terminology, a callback invocation is the result of an event that reflects something happening on the front-end. A variable was changed (`on_change`), a button was pressed (`on_action`) and so on.
This is really not the topic here. And again, the original wording of the issue made it confusing. I apologize for that.

- The proposal to create a method of Gui makes things clear... but wrong, in my opinion.
  The functionality is not related to any Gui. A Taipy GUI application with several Gui instances (although we're still wondering what the use case would be) would not really know what to do.

- The role of the user-defined function is to convert data with an unknown type into a typed value that Taipy supports.
  That could be seen as a "type converter," but the naming is confusing since what it does is really convert the data, with the type change being a (desired) side effect.
  So let's call this a "data converter". Actually, that's not so bad, since we can implement whatever we want, right?
  It's a shame there's so little context when the function is invoked (then the conversion could benefit from more information to do things differently), but we have what we have.
  And what we want is to handle values with unknown types. Not so much invalid. Now we get to "unknown data converter".
  And this would be exposed from the *data* module. Finally, "unknown converter" can do the job.
  I would certainly create a global function (`taipy.gui.data.set_unknown_converter()`) that can be invoked with any user-defined function expecting any value (knowing that its type was not recognized by Taipy for serialization) and returning any other value with a type that Taipy does supports.
  A single global variable (similar to current's `Gui.on_invalid_data`) will privately keep this from the user's eyes.

The bottom line is this exposed API:

```
# set_unknown_converter() would be defined in or imported by taipy/gui/data/__init__.py

def my_amazing_converter(value: any) -> Any:
    return ... # Whatever transforms value into something Taipy can use
  
taipy.gui.data.set_unknown_converter(my_amazing_converter)
```
  
It is implemented in a manner very similar to what you have proposed, and I volunteer to refactor all this.
Of course @Ujj1225 the 100 pts are yours.

@FredLL-Avaiga what say you?

**Comment by FabienLelaquais** [comment]
@Ujj1225 are you still there?

**Comment by Ujj1225** [comment]
yes @FabienLelaquais 
I am keen to see on how you are shaping your approach with this one. Waiting patiently. 
Please let me know if i can be of some help

**Comment by FabienLelaquais** [comment]
> yes @FabienLelaquais I am keen to see on how you are shaping your approach with this one. Waiting patiently. Please let me know if i can be of some help

Sure thing. I was for some reason waiting for your approval, but a look at my message clearly shows I should not have.
I'll take care of the rest of this. Thanks again for your help, @Ujj1225! Take care.

File: tests/gui/gui_specific/test_get_instance.py (hunk 15 above)
Comment (FabienLelaquais) [reply tests/gui/gui_specific/test_get_instance.py]: Done

**Comment by FabienLelaquais** [comment]
@Ujj1225, @FredLL-Avaiga 
As promised I handled the refactoring if the original code.
Not much I admit, but  like it better.
Please let me know what you think.
Cheers

File: taipy/gui/gui.py [hunk 18]
```diff
@@ -2637,6 +2672,7 @@ def __bind_default_function(self):
             self.__bind_local_func("on_exception")
             self.__bind_local_func("on_status")
             self.__bind_local_func("on_user_content")
+            self.__bind_local_func("on_invalid_data")
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: I forgot to remove this

File: tests/gui/data/test_accessors.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+    # Testing when accessor is not registered
+    data_accessor = data_accessors._DataAccessors__get_instance(mock_taipy_data)  # type: ignore
+    assert isinstance(
+        data_accessor, _InvalidDataAccessor
+    ), f"Expected _InvalidDataAccessor but got {type(data_accessor)}"
+    result = data_accessors.get_data("var_name", data, {})
+    assert result == {}
+
+    # Testing when accessor is not registered
```
Comment (FabienLelaquais) [inline tests/gui/data/test_accessors.py]:
```suggestion
    # Testing when accessor is registered
```

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Details - but.

File: contributors.txt
Comment (FredLL-Avaiga) [inline contributors.txt:1]: how come ? has this file moved ?

File: taipy/gui/data/data_accessor.py [hunk 20]
```diff
@@ -132,11 +133,21 @@ def _register(self, cls: t.Type[_DataAccessor]) -> None:
...
+        """Unregisters a DataAccessor type."""
+        if cls in self.__access_4_type:
+            del self.__access_4_type[cls]
+
     def __get_instance(self, value: _TaipyData) -> _DataAccessor:  # type: ignore
         value = value.get() if isinstance(value, _TaipyData) else value
         access = self.__access_4_type.get(type(value))
         if access is None:
             if value is not None:
+                converted_value = type(self.__gui)._convert_unsupported_data(value)
```
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py:146]: is the `type(self.__gui)` needed ?
I would think that `self.__gui.._convert_unsupported_data(value)` would work ?

File: taipy/gui/gui.py [hunk 21]
```diff
@@ -1143,6 +1143,41 @@ def __update_state_context(self, payload: dict):
             for var, val in state_context.items():
                 self._update_var(var, val, True, forward=False)
 
+
+    @staticmethod
+    def set_unsupported_data_converter(converter: t.Optional[t.Callable[[t.Any], t.Any]]) -> None:
+        """Set a custom converter for unsupported data types.
+
+        This function allows specifying a custom conversion function for data types that cannot
+        be serialized automatically. When Taipy GUI encounters an unsupported type in the data
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py:1152]: this is invoked only when the property is types as TaipyData

File: taipy/gui/gui.py [hunk 22]
```diff
@@ -1143,6 +1143,41 @@ def __update_state_context(self, payload: dict):
...
+        Handles unsupported data by invoking the converter if there is one.
+
+        Arguments:
+            value: The unsupported data encountered.
+
+        Returns:
+            The transformed data or None if no transformation is possible.
+        """
+        try:
+            return Gui.__unsupported_data_converter(value) if callable(Gui.__unsupported_data_converter) else None  # type: ignore
```
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: use `_is_function` instead of `callable`

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

File: taipy/gui/data/data_accessor.py (hunk 20 above)
Comment (FabienLelaquais) [reply taipy/gui/data/data_accessor.py:146]: `_convert_unsupported_data()` is a class method.

File: taipy/gui/gui.py (hunk 21 above)
Comment (FabienLelaquais) [reply taipy/gui/gui.py:1152]: I'll add this... when TaipyData is documented.

File: taipy/gui/gui.py (hunk 22 above)
Comment (FabienLelaquais) [reply taipy/gui/gui.py]: I would recommend not to. 
A class (with `__call__()` is to be a valid value here.
Tell me.

File: contributors.txt
Comment (FabienLelaquais) [reply contributors.txt:1]: This should no have been created indeed

File: taipy/gui/gui.py (hunk 22 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: I thought `_is_function` was created for this :-(

Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: but a class without `__call__` would be `callable`

Comment (FabienLelaquais) [reply taipy/gui/gui.py]: Whaat?

Comment (FabienLelaquais) [reply taipy/gui/gui.py]: Oh I see. `MyClass` class is always callable since you can write `MyClass()`... which invokes the constructor.

File: taipy/gui/data/data_accessor.py (hunk 20 above)
Comment (FredLL-Avaiga) [reply taipy/gui/data/data_accessor.py:146]: I get that but it can be called on the instance, no ?

File: contributors.txt
Comment (FredLL-Avaiga) [inline contributors.txt:1]: I don't get why this file is completely changed ?

**Review by FredLL-Avaiga — DISMISSED** [review]

File: contributors.txt
Comment (FabienLelaquais) [reply contributors.txt:1]: My bad - this should definitively go away. We want Ujj1225 in that's all.

Comment (FabienLelaquais) [reply contributors.txt:1]: This file used to be there, then was moved elsewhere. I'll check this out before I merge.
Thanks!

File: taipy/gui/data/data_accessor.py (hunk 20 above)
Comment (FabienLelaquais) [reply taipy/gui/data/data_accessor.py:146]: No it cannot. `<class>.method()` is different from `<instance>.method()`.

**Review by FredLL-Avaiga — APPROVED** [review]

**Review by FabienLelaquais — APPROVED** [review]
